core/src/main/java/hudson/model/Queue.java

```diff
@@ -377,6 +396,9 @@ public synchronized void load() {
                             if(item.task==null)
                                 continue;   // botched persistence. throw this one away
 
+                            // backward compatibility - in case the serialized queue state was
+                            // a List of items (as of old), then the maxId is the max of the items in
+                            // the persisted queue.
```

Comment:
For the compatibility loop you can look for `Task`, scheduling it; else look for `Item`, updating `maxId` and also adding it to the list of items to process. So

    List<Item> items;

is what I should have written.